File: leetcode2/1easy/최원준/Q1592.java
package Leetcode.최원준;

/*
1. 아이디어 :


2. 시간복잡도 :
O( )

3. 자료구조/알고리즘 :

 */

import java.util.ArrayList;
import java.util.List;

public class Q1592 {
    class Solution {
        public String reorderSpaces(String text) {
            int count = 0;
            boolean prevChar = true;

            List<String> words = new ArrayList<>();
            StringBuilder sb = new StringBuilder();

            for (char c : text.toCharArray()) {
                if (c == ' ') {
                    if (prevChar) {
                        prevChar = false;
                        if (!sb.isEmpty()) {
                            words.add(sb.toString());
                            sb = new StringBuilder();
                        }

                    }
                    count++;
                } else {
                    sb.append(c);
                    prevChar = true;
                }
            }
            if (!sb.isEmpty()) words.add(sb.toString());

            if (words.size()==1) {
                return words.get(0) + " ".repeat(count);
            }

            int equalGaps = count / (words.size()-1);
            int remain = count % (words.size()-1);

            sb = new StringBuilder();
            for (int i=0; i<words.size()-1; i++) {
                sb.append(words.get(i));
                sb.append(" ".repeat(equalGaps));
            }
            sb.append(words.get(words.size()-1)).append(" ".repeat(remain));

            return sb.toString();
        }
    }
}

File: leetcode2/1easy/노형민/Q1592.cs
using System.Text;

public class Solution
{
    public string ReorderSpaces(string text)
    {
        StringBuilder sb = new();
        int spaceCount = 0;
        int wordCount = 0;
        bool inWord = false;

        foreach (char c in text)
        {
            if (c == ' ')
            {
                spaceCount++;
                inWord = false;
            }
            else
            {
                if (!inWord)
                {
                    wordCount++;
                    inWord = true;
                }
            }
        }

        if (wordCount != 1)
        {
            bool wordStarted = true;
            bool firstWord = true;
            foreach (char c in text)
            {
                if (c == ' ')
                {
                    wordStarted = false;
                }
                else
                {
                    if (!wordStarted && !firstWord)
                    {
                        sb.Append(' ', spaceCount / (wordCount - 1));
                    }

                    sb.Append(c);
                    wordStarted = true;
                    firstWord = false;
                }
            }
            
            sb.Append(' ', spaceCount % (wordCount - 1));
            
            return sb.ToString();
        }
        else
        {
            sb.Append(text.Trim());
            sb.Append(' ', spaceCount);
            return sb.ToString();
        }
    }
}